src/pipdeptree/_models/package.py

```diff
@@ -23,13 +23,14 @@ class Package(ABC):
 
     UNKNOWN_LICENSE_STR = "(Unknown license)"
 
-    def __init__(self, obj: DistInfoDistribution) -> None:
-        self._obj: DistInfoDistribution = obj
-        self.key = pep503_normalize(obj.key)
+    def __init__(self, obj: Distribution) -> None:
+        self._obj: Distribution = obj
+        self.key = pep503_normalize(obj.name)
```

Comment:
As I mentioned before, `Distribution.name()` was added in 3.10 and on. Let's just use `Distribution.metadata['Name']` here instead.

Reply:
`Requirement` in `packaging` has no attribute 'metadata' which means `super().__init__(obj)` may failed. So I think we should make its own `key` without `super().__init__(obj)`?

Reply:
Sorry I did this review late into the night; since `Package` uses `key` in some of its methods, we can do this is:

- Add a `def key(): ...` property to `Package` that is an abstract method
- Have both `DistPackage` and `ReqPackage` implement this `key` property

There are better ways but this approach has us avoid breaking everyone who is using `Package`'s constructor right now.

Reply:
Ok, thank you, this is truely a better way. I have done this part and `project_name` too.

Reply:
And I found that `key` and `project_name` of Class `Package` look a little bit duplicated. One of them is some kind enough I think.

Reply:
This is expected. The problem I found is if we change __init__'s signature we will be breaking everyone that calls it which will lead to many areas in the code that need changing. I don't want to worry about in this PR.

Reply:
This is in a `_models` package, so is not public so no one should be calling it, no?

Reply:
> This is in a `_models` package, so is not public so no one should be calling it, no?

Actually I just meant any of our code that is calling on `Package`'s `__init__()`.  I thought we had many callers of this constructor but looking into this further I realized that the only ones who are calling this constructor are just `DistPackage` and `ReqPackage` in their constructors.

@xiacunshun With this I feel like your idea of adding `key` and `project_name` to `Package`'s constructor is much better so let's do that.

Reply:
No problem, let's complete this PR better. For this purpose, I can keep improving the code until it meets the requirements.

Reply:
Done this.